packages/cojson-storage-indexeddb/src/idbClient.ts

```diff
@@ -0,0 +1,234 @@
+        };
+      }) => {
+        const request = handler(stores);
+        request.onerror = () => {
+          console.error("Error in request", request.error);
+          this.currentTx = undefined;
+          reject(request.error);
+          // TODO: recover pending requests in new tx
+        };
+        request.onsuccess = () => {
+          const value = request.result as T;
+          resolve(value);
+
+          const next = txEntry!.pendingRequests.shift();
+
+          if (next) {
+            next({ stores });
+          } else {
+            if (this.currentTx === txEntry) {
+              this.currentTx = undefined;
+            }
+          }
+        };
+      };
+
+      if (!txEntry || performance.now() - txEntry.startedAt > 20) {
+        const tx = this.db.transaction(
+          ["coValues", "sessions", "transactions", "signatureAfter"],
+          "readwrite",
+        );
+        txEntry = {
+          id: this.currentTxID++,
+          tx,
+          stores: {
+            coValues: tx.objectStore("coValues"),
+            sessions: tx.objectStore("sessions"),
+            transactions: tx.objectStore("transactions"),
+            signatureAfter: tx.objectStore("signatureAfter"),
+          },
+          startedAt: performance.now(),
+          pendingRequests: [],
+        };
+
+        this.currentTx = txEntry;
+
+        requestEntry(txEntry);
+      } else {
+        txEntry.pendingRequests.push(requestEntry);
+      }
+    });
+  }
+
+  async getCoValue(coValueId: RawCoID): Promise<StoredCoValueRow | undefined> {
+    return this.makeRequest<StoredCoValueRow | undefined>(({ coValues }) =>
+      coValues.index("coValuesById").get(coValueId),
+    );
+  }
+
+  async getCoValueSessions(coValueRowId: number): Promise<StoredSessionRow[]> {
+    return this.makeRequest<StoredSessionRow[]>(({ sessions }) =>
+      sessions.index("sessionsByCoValue").getAll(coValueRowId),
+    );
+  }
+
+  async getNewTransactionInSession(
+    sessionRow: StoredSessionRow,
+    firstNewTxIdx: number,
+  ): Promise<TransactionRow[]> {
+    return this.makeRequest<TransactionRow[]>(({ transactions }) =>
+      transactions.getAll(
+        IDBKeyRange.bound(
+          [sessionRow.rowID, firstNewTxIdx],
+          [sessionRow.rowID, Infinity],
+        ),
+      ),
+    );
+  }
+
+  async getSignatures(
+    sessionRow: StoredSessionRow,
+    firstNewTxIdx: number,
+  ): Promise<SignatureAfterRow[]> {
+    return this.makeRequest<SignatureAfterRow[]>(
+      ({ signatureAfter }: { signatureAfter: IDBObjectStore }) =>
+        signatureAfter.getAll(
+          IDBKeyRange.bound(
+            [sessionRow.rowID, firstNewTxIdx],
+            [sessionRow.rowID, Infinity],
+          ),
+        ),
+    );
+  }
+
+  async addCoValue(
+    msg: CojsonInternalTypes.NewContentMessage,
+  ): Promise<number> {
+    return (await this.makeRequest<IDBValidKey>(({ coValues }) =>
+      coValues.put({
+        id: msg.id,
+        header: msg.header!,
```

Comment:
Header is of type CoValueHeader|undefined and we save it as it is (no boolean).
This code l.182 had to narrow types as I thought 
```
   if (!msg.header) {
      throw new Error("Header is required, coId: " + msg.id);
    }
    ```
But TS still says  it could be undefined so I had to make it explicit `msg.header!`